app/services/sign_in/refresh_token_decryptor.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+module SignIn
+  class RefreshTokenDecryptor
+    attr_reader :encrypted_refresh_token
+
+    def initialize(encrypted_refresh_token:)
+      @encrypted_refresh_token = encrypted_refresh_token
+    end
+
+    def perform
+      split_refresh_token = split_encrypted_refresh_token(encrypted_refresh_token)
+
+      decrypted_part = get_decrypted_part(split_refresh_token)
+      nonce = get_nonce_part(split_refresh_token)
+      version = get_version_part(split_refresh_token)
+
+      validate_token!(decrypted_part, nonce, version)
+
+      SignIn::RefreshToken.new(session_handle: decrypted_part.session_handle,
+                               user_uuid: decrypted_part.user_uuid,
+                               parent_refresh_token_hash: decrypted_part.parent_refresh_token_hash,
+                               anti_csrf_token: decrypted_part.anti_csrf_token,
+                               nonce: decrypted_part.nonce,
+                               version: decrypted_part.version)
+    end
+
+    private
+
+    def validate_token!(decrypted_token, nonce, version)
+      raise SignIn::Errors::RefreshVersionMismatchError if decrypted_token.version != version
+      raise SignIn::Errors::RefreshNonceMismatchError if decrypted_token.nonce != nonce
+    end
+
+    def get_decrypted_part(split_refresh_token)
+      decrypt_refresh_token(split_refresh_token[SignIn::Constants::REFRESH_TOKEN_ENCRYPTED_POSITION])
+    end
+
+    def get_nonce_part(split_refresh_token)
+      split_refresh_token[SignIn::Constants::REFRESH_TOKEN_NONCE_POSITION]
+    end
+
+    def get_version_part(split_refresh_token)
+      split_refresh_token[SignIn::Constants::REFRESH_TOKEN_VERSION_POSITION]
+    end
+
+    def split_encrypted_refresh_token(encrypted_refresh_token)
+      encrypted_refresh_token.split('.')
+    end
+
+    def decrypt_refresh_token(encrypted_refresh_token)
+      message_encryptor.decrypt_and_verify(encrypted_refresh_token)
+    end
+
+    def message_encryptor
+      ActiveSupport::MessageEncryptor.new(Rails.application.secrets.secret_key_base[0..31])
```

Comment:
Should we be using Lockbox for encryption here so that the key gets rotated appropriately?

Reply:
Updated this to use KMSEncrypted for encrypting/decrypting